grails-core/src/main/groovy/org/grails/config/yaml/YamlPropertySourceLoader.groovy

```diff
@@ -42,61 +44,72 @@ class YamlPropertySourceLoader extends YamlProcessor implements PropertySourceLo
-            if(matchDefault && resource.filename == Metadata.FILE) {
+            if (matchDefault && resource.filename == Metadata.FILE) {
                 def metadata = Metadata.getCurrent()
                 def metadataSource = metadata.getSource()
                 def metadataFile = metadata.getMetadataFile()
-                if(metadataSource != null && metadataFile != null && metadataFile.getURL().equals(resource.getURL())) {
-                    for(o in metadataSource) {
-                        if(o instanceof Map) {
-                            propertySource.merge((Map)o, false)
+                if (metadataSource != null && metadataFile != null && metadataFile.getURL().equals(resource.getURL())) {
+                    for (o in metadataSource) {
+                        if (o instanceof Map) {
+                            propertySource.merge((Map) o, false)
                         }
                     }
-                }
-                else {
+                } else {
                     process { Properties properties, Map<String, Object> map ->
                         propertySource.merge(map, parseFlatKeys)
                     }
                 }
-            }
-            else {
+            } else {
                 process { Properties properties, Map<String, Object> map ->
-                    if(!filteredKeys.isEmpty()) {
+                    if (!filteredKeys.isEmpty()) {
                         def env = map.get(GrailsPlugin.ENVIRONMENTS)
-                        for(key in filteredKeys) {
+                        for (key in filteredKeys) {
                             map.remove(key)
-                            if(env instanceof Map) {
-                                Map envMap = (Map)env
-                                for(envSpecific in envMap.values()) {
-                                    if(envSpecific instanceof Map) {
-                                        ((Map)envSpecific).remove(key)
+                            if (env instanceof Map) {
+                                Map envMap = (Map) env
+                                for (envSpecific in envMap.values()) {
+                                    if (envSpecific instanceof Map) {
+                                        ((Map) envSpecific).remove(key)
                                     }
                                 }
                             }
                         }
                     }
+
+                    //Now merge the environment config over the top of the normal stuff
+                    def environments = map.get(GrailsPlugin.ENVIRONMENTS)
+                    if (environments instanceof Map) {
+                        Map envMap = (Map) environments
+                        for (envSpecific in envMap) {
+                            if (envSpecific instanceof Map.Entry) {
+                                if (envSpecific?.key?.toString()?.equalsIgnoreCase(Environment.getCurrentEnvironment()?.name)) {
```

Comment:
I didn't combine due to the first loops dependence on the exclude list that should not be empty, but I can ref actor them together.